File: internal/scraper/cgbookcase.go
package scraper

import (
	"context"
	"log/slog"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/struffel/3d-assets-one/internal/model"
)

type CgBookcaseScraper struct{}

func (s *CgBookcaseScraper) Creator() model.Creator { return model.CreatorCGBookcase }

func (s *CgBookcaseScraper) ScrapeAssets(ctx context.Context, existing []model.StoredAsset) ([]ScrapedAsset, error) {
	const maxPerRun = 5
	const baseURL = "https://www.cgbookcase.com/textures/"

	body, _, err := FetchURL(ctx, baseURL)
	if err != nil {
		return nil, err
	}

	doc, err := goquery.NewDocumentFromReader(strings.NewReader(body))
	if err != nil {
		return nil, err
	}

	var urls []string
	doc.Find(`a[href*="/textures/"]`).Each(func(_ int, sel *goquery.Selection) {
		href, exists := sel.Attr("href")
		if exists {
			urls = append(urls, "https://www.cgbookcase.com"+href+"?source=3dassets.one")
		}
	})

	var results []ScrapedAsset
	for _, u := range urls {
		if len(results) >= maxPerRun {
			break
		}
		if ContainsURL(existing, u) {
			continue
		}

		meta, err := parseHTMLMetaTags(ctx, u)
		if err != nil {
			continue
		}

		name, ok := meta["tex1:name"]
		if !ok || name == "" {
			continue
		}
		previewURL, ok := meta["tex1:preview-image"]
		if !ok || previewURL == "" {
			continue
		}
		tagStr, ok := meta["tex1:tags"]
		if !ok {
			continue
		}
		typeStr := meta["tex1:type"]

		img, err := FetchImage(ctx, previewURL)
		if err != nil {
			slog.Warn("cgbookcase: thumbnail failed", "url", u, "err", err)
			continue
		}

		results = append(results, ScrapedAsset{
			Title:        name,
			URL:          u,
			Tags:         ExplodeFilterTrim(",", tagStr),
			Type:         model.AssetTypeFromTex1Tag(typeStr),
			Creator:      model.CreatorCGBookcase,
			Status:       model.ScrapedNewlyFound,
			RawThumbnail: img,
		})
	}

	return results, nil
}

func (s *CgBookcaseScraper) ValidateAsset(ctx context.Context, asset model.StoredAsset) bool {
	return HeadOK(ctx, asset.URL)
}

File: app/creator/logic/cgbookcase.py
from app.asset.asset_type import AssetType
from app.asset.scraped_asset import ScrapedAsset
from app.asset.scraped_asset_collection import ScrapedAssetCollection
from app.asset.scraped_asset_status import ScrapedAssetStatus
from app.asset.stored_asset_collection import StoredAssetCollection
from app.creator.creator import Creator
from app.creator.creator_logic import CreatorLogic
from app.fetch.web_item_reference import WebItemReference
from app.misc.string_util import StringUtil


class CreatorLogicCgBookcase(CreatorLogic):
    creator = Creator.CGBOOKCASE
    max_assets_per_run = 5
    base_url = "https://www.cgbookcase.com/textures/"

    def scrape_assets(self, existing_assets: StoredAssetCollection) -> ScrapedAssetCollection:
        dom = WebItemReference(self.base_url).fetch().parse_as_dom()
        if dom is None:
            return ScrapedAssetCollection()

        asset_links = dom.find_all("a", href=lambda h: h and "/textures/" in h)
        url_array = [
            "https://www.cgbookcase.com" + a.get("href", "") + "?source=3dassets.one"
            for a in asset_links
            if a.get("href")
        ]

        tmp_collection = ScrapedAssetCollection()
        count = 0

        for url in url_array:
            if existing_assets.contains_url(url):
                continue

            meta = WebItemReference(url).fetch().parse_html_meta_tags()
            if meta is None:
                continue
            required = {"tex1:name", "tex1:release-date", "tex1:tags", "tex1:type", "tex1:preview-image"}
            if not required.issubset(meta.keys()):
                continue

            asset = ScrapedAsset(
                id=None,
                creator_given_id=None,
                title=meta["tex1:name"],
                url=url,
                tags=StringUtil.explode_filter_trim(",", meta["tex1:tags"]),
                type=AssetType.from_tex1_tag(meta["tex1:type"]),
                creator=Creator.CGBOOKCASE,
                status=ScrapedAssetStatus.NEWLY_FOUND,
                raw_thumbnail=WebItemReference(url=meta["tex1:preview-image"]).fetch().parse_as_image(),
            )
            tmp_collection.append(asset)
            count += 1
            if count >= self.max_assets_per_run:
                break

        return tmp_collection
